src/ansi-c/ansi_c_internal_additions.cpp

```diff
@@ -120,6 +122,29 @@ static std::string architecture_string(T value, const char *s)
          std::string(s) + "=" + std::to_string(value) + ";\n";
 }
 
+/// The maximum allocation size is determined by the number of bits that
+/// are left in the pointer of width \p pointer_width.
+///
+/// The offset needs to be able to represent up to the allocation_size, and
+/// down to -allocation_size, hence 2^(pointer_width - object_bits - 1)
+/// \param pointer_width: The width of the pointer
+/// \param object_bits : The number of bits used to represent the ID
+/// \return The size in bytes of the maximum allocation supported.
+static size_t
+max_malloc_size(std::size_t pointer_width, std::size_t object_bits)
+{
+  PRECONDITION(pointer_width >= 1);
+  PRECONDITION(object_bits < pointer_width);
+  PRECONDITION(object_bits >= 1);
+  const auto offset_bits = pointer_width - object_bits;
+  // We require the offset to be able to express upto allocation_size - 1,
+  // but also down to -allocation_size + 1, therefore the size is allowable
```

Comment:
Should this be just `-allocation_size`? Since we can represent one more negative value than positive values.

Reply:
Yes you're right - a pointer to the end of the memory would be valid, and you'd need to subtract the full allocation size to get back to the start.